compiler/rustc_query_system/src/query/job.rs

```diff
@@ -478,36 +478,40 @@ fn remove_cycle(
 /// If a query cycle is found it will break the cycle by finding an edge which
 /// uses a query latch and then resuming that waiter.
 /// There may be multiple cycles involved in a deadlock, so this searches
-/// all active queries for cycles before finally resuming all the waiters at once.
+/// all active queries for cycles. But only one waiter will be resumed at once.
 pub fn break_query_cycles(query_map: QueryMap, registry: &rayon_core::Registry) {
-    let mut wakelist = Vec::new();
+    // FIXME: change `remove_cycle` to make it return one waiter at a call.
+    // So that we can avoid use the global list here.
+    static WAKELIST: Mutex<Vec<Arc<QueryWaiter>>> = Mutex::new(Vec::new());
+
     let mut jobs: Vec<QueryJobId> = query_map.keys().cloned().collect();
 
     let mut found_cycle = false;
 
     while jobs.len() > 0 {
-        if remove_cycle(&query_map, &mut jobs, &mut wakelist) {
+        if remove_cycle(&query_map, &mut jobs, &WAKELIST) {
             found_cycle = true;
```

Comment:
I'd try to put a break below `found_cycle = true;`, with a `FIXME` to remove it later, once the real cause of the deadlocks is fixed. All other changes in the PR could be removed.

Reply:
Ah, it sounds good. So `remove_cycle` doesn't always return `true` when there are deadlocks

Reply:
Thanks! Has done it